sim/src/bookinfo/leafnode.rs

```diff
@@ -0,0 +1,219 @@
+//! An abstraction of a node.  The node can have a plugin, which is meant to reprsent a WebAssembly filter
+//! A node is a sim_element.
+
+use crate::plugin_wrapper::PluginWrapper;
+use crate::sim_element::SimElement;
+use core::any::Any;
+use queues::*;
+use rpc_lib::rpc::Rpc;
+use std::cmp::min;
+use std::fmt;
+
+#[derive(Clone)]
+pub struct RpcWithSender {
+    pub rpc: Rpc,
+    pub sender: String,
+}
+pub struct LeafNode {
+    queue: Queue<RpcWithSender>,   // queue of rpcs
+    id: String,                    // id of the node
+    capacity: u32,                 // capacity of the node;  how much it can hold at once
+    egress_rate: u32,              // rate at which the node can send out rpcs
+    generation_rate: u32, // rate at which the node can generate rpcs, which are generated regardless of input to the node
+    plugin: Option<PluginWrapper>, // filter to the node
+    neighbors: Vec<String>, // who is the node connected to
+}
+
+impl fmt::Display for LeafNode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(width) = f.width() {
+            if self.plugin.is_none() {
+                write!(f, "{:width$}",
+                       &format!("LeafNode {{ id : {}, capacity : {}, egress_rate : {}, generation_rate : {}, queue: {}, plugin : None}}",
+                       &self.id, &self.capacity, &self.egress_rate, &self.generation_rate, &self.queue.size()),
+                       width = width)
+            } else {
+                write!(f, "{:width$}",
+                       &format!("LeafNode {{ id : {}, capacity : {}, egress_rate : {}, generation_rate : {}, queue : {}, \n\tplugin : {} }}",
+                       &self.id, &self.capacity, &self.egress_rate, &self.generation_rate, &self.queue.size(), self.plugin.as_ref().unwrap()),
+                       width = width)
+            }
+        } else {
+            if self.plugin.is_none() {
+                write!(f, "LeafNode {{ id : {}, egress_rate : {}, generation_rate : {}, plugin : None, capacity : {}, queue : {} }}",
+                       &self.id, &self.egress_rate, &self.generation_rate, &self.capacity, &self.queue.size())
+            } else {
+                write!(
+                    f,
+                    "LeafNode {{ id : {}, egress_rate : {}, generation_rate : {}, plugin : {}, capacity : {}, queue : {} }}",
+                    &self.id,
+                    &self.egress_rate,
+                    &self.generation_rate,
+                    self.plugin.as_ref().unwrap(),
+                    &self.capacity,
+                    &self.queue.size()
+                )
+            }
+        }
+    }
+}
+
+impl SimElement for LeafNode {
+    fn tick(&mut self, tick: u64) -> Vec<(Rpc, String)> {
+        let mut ret = vec![];
```

Comment:
Considering that this is a leaf node, we can simplify this a lot. Right?

Reply:
Simplify it how?  It seems like it's doing the normal things for a leaf node:  producing RPCs according to its generation rate, and sending any request RPCs back to the sender as responses.

Reply:
Yeah lets just leave it at that for now.